gson/src/test/java/com/google/gson/functional/JsonOrgInteropTest.java

```diff
@@ -0,0 +1,589 @@
+ * Tests interoperability with https://github.com/stleary/JSON-java ({@code org.json} package).
+ */
+public class JsonOrgInteropTest {
+  @Test
+  public void testNoCustomAdapter() {
+    Gson gson = new Gson();
+    String expectedMessageArray = "Unsupported class from other JSON library: org.json.JSONArray"
+        + "\nSee https://github.com/google/gson/blob/main/Troubleshooting.md#unsupported-json-library-class";
+    String expectedMessageObject = "Unsupported class from other JSON library: org.json.JSONObject"
+        + "\nSee https://github.com/google/gson/blob/main/Troubleshooting.md#unsupported-json-library-class";
+
+    // TODO: Adjust these once more specific exception type than RuntimeException is thrown
+    Exception e = assertThrows(RuntimeException.class, () -> gson.toJson(new JSONArray()));
+    assertThat(e).hasMessageThat().isEqualTo(expectedMessageArray);
+
+    e = assertThrows(RuntimeException.class, () -> gson.toJson(new JSONObject()));
+    assertThat(e).hasMessageThat().isEqualTo(expectedMessageObject);
+
+    e = assertThrows(RuntimeException.class, () -> gson.fromJson("[]", JSONArray.class));
+    assertThat(e).hasMessageThat().isEqualTo(expectedMessageArray);
+
+    e = assertThrows(RuntimeException.class, () -> gson.fromJson("{}", JSONObject.class));
+    assertThat(e).hasMessageThat().isEqualTo(expectedMessageObject);
+  }
+
+  // Custom classes for equality assertions to avoid using directly JSONArray and JSONObject
+  // which perform element wrapping and conversion, and because their `toList()` and `toMap()`
+  // methods also recursively convert values
+  private static class ExpectedJSONArray {
+    public final List<Object> elements;
+
+    public ExpectedJSONArray(List<Object> elements) {
+      this.elements = elements;
+    }
+
+    @Override
+    public String toString() {
+      return "JSONArray" + elements;
+    }
+  }
+
+  private static class ExpectedJSONObject {
+    public final Map<String, Object> entries;
+
+    public ExpectedJSONObject(Map<String, Object> entries) {
+      this.entries = entries;
+    }
+
+    @Override
+    public String toString() {
+      return "JSONObject" + entries;
+    }
+  }
+
+  private abstract static class JsonOrgBaseSubject extends Subject {
+
+    protected JsonOrgBaseSubject(FailureMetadata metadata, @Nullable Object actual) {
+      super(metadata, actual);
+    }
+
+    protected void checkElementValues(String message, Object expected, Object actual) {
+      StandardSubjectBuilder builder = check(message);
+
+      if (actual instanceof JSONArray) {
+        builder.about(JSONArraySubject.jsonArrays()).that((JSONArray) actual).isEqualTo(expected);
+      } else if (actual instanceof JSONObject) {
+        builder.about(JSONObjectSubject.jsonObjects()).that((JSONObject) actual).isEqualTo(expected);
+      } else {
+        builder.that(actual).isEqualTo(expected);
+      }
+    }
+  }
+
+  private static class JSONArraySubject extends JsonOrgBaseSubject {
```

Comment:
Maybe having these custom Truth `Subject` classes is a bit exaggerated, and they might not follow Truth best practices (feedback is welcome). Though on the other hand they make comparison of `JSONArray` and `JSONObject` easier.